modules/gem5/_posts/2000-04-01-assignment2.md

```diff
@@ -0,0 +1,247 @@
+
+This homework is based on exercise 3.6 of [CA:AQA 3rd edition](https://www.google.com/books/edition/Computer_Architecture/XX69oNsazH4C) (the former textbook for this course) and was developed in part by Jason Lowe-Power et al., then modernized by Matt Sinclair and Jason Lowe-Power.
+
+## Workload
+
+For this assignment we are going to use DAXPY as our workload. The DAXPY loop (double precision `aX + Y`) is an often used operation in programs that work with matrices and vectors. The following code implements DAXPY in C++.
+
+```cpp
+#include <cstdio>
+#include <random>
+
+int main()
+{
+  const int N = 4096;
+  double X[N], Y[N], alpha = 0.5;
+  std::random_device rd; std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dis(1, 2);
+  for (int i = 0; i < N; ++i)
+  {
+    X[i] = dis(gen);
+    Y[i] = dis(gen);
+  }
+
+  // Start of daxpy loop
+  for (int i = 0; i < N; ++i)
+  {
+    Y[i] = alpha * X[i] + Y[i];
+  }
+  // End of daxpy loop
+
+  double sum = 0;
+  for (int i = 0; i < N; ++i)
+  {
+    sum += Y[i];
+  }
+  printf("%lf\n", sum);
+  return 0;
+}
+```
+
+You can find the definitions for the workload objects in gem5 under `workloads/workloads.py`. In this assignment, we will only be using `DAXPYWorkload`. In order to create an object of `DAXPYWorkload` you just need to call its constructor (`__init__`) function.
+
+## Experimental setup
+
+In this assignment we are going to measure the impact of different pipeline latencies on the overall performance of the system.
+As we discussed in the class it is important to consider measured performance as a product of both software and hardware.
+In this spirit, it might be useful to get a picture of the instruction mix in our workload.
+You might find this information useful in later steps of your analysis.
+As part of this assignment, you will only modify/change the CPU model.
+Models for the board, cache hierarchy, and memory will remain a constant in your experiment.
+
+- Board models: You can find all the models you need to use for your CPU (processor) under `components/boards.py`. You will only be using `HW2RISCVBoard` in this assignment.
+- CPU models: You can find all the models you need to use for your CPU (processor) under `components/processors.py`.
+- Cache models: You can find all the models you need to use for your cache hierarchy under `components/cache_hierarchies.py`. You will only use `HW2MESITwoLevelCache` in this assignment.
+- Memory models: You can find all the models you need to use for your memory under `components/memories.py`. You will only use `HW2DDR3_1600_8x8` in this assignment.
+
+### **IMPORTANT NOTE**
+
+In your configuration scripts, make sure to import `exit_event_handler` using the command below.
+
+```python
+from workloads.roi_manager import exit_event_handler
+```
+
+You will have to pass `exit_event_handler` as a keyword argument named `on_exit_event` when creating a `simulator` object. Use the *template* below to create a simulator object.
```

Comment:
It may be worth adding a couple of sentences of context here to explain what an exit event handler is and why it matters.